### PR TITLE
New KleidiAI basics Learning Path

### PR DESCRIPTION
**Pull request by zachlasiuk**
Before submitting a pull request for a new Learning Path, please review [Create a Learning Path](https://learn.arm.com//learning-paths/cross-platform/_example-learning-path/)
- [x] I have reviewed Create a Learning Path

Please do not include any confidential information in your contribution. This includes confidential microarchitecture details and unannounced product information. No AI tool can be used to generate either content or code when creating a learning path or install guide.

- [x] I have checked my contribution for confidential information

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of the Creative Commons Attribution 4.0 International License.

### CODE REVIEW

File: content/learning-paths/cross-platform/kleidiai-explainer/page3.md [hunk 1]
```diff
@@ -0,0 +1,256 @@
...
+The example code is structured to take in two matricies and compare KleidiAI's micro-kernel results to a reference implementation. This reference code is functionally identical to KleidiAI's micro-kernels, and is present for two reasons:
+1. To check for KleidiAI micro-kernel output validity as a sanity check, and
+2. To explain how KleidiAI micro-kernels work. The implementation within KleidiAI's micro-kernels use hand-optimized C and Arm Assembly code, naturally being difficult to interpret.
+
+
+## Build and Run the KleidiAI C++ example (i8mm micro-kernel)
+
+Follow these steps to build and run the KleidiAI library and example script reviewing the:
+
+1. Create an Ubuntu 24 Arm Linux machine on an AWS EC2 instance. Use an M7g-medium, which uses the Graviton 3 SoC supporting the *i8mm* Arm architecture feature. The 1 CPU and 4 GB of RAM in the M7g-medium are sufficient for this basic example run.
```
Comment (annietllnd) [inline content/learning-paths/cross-platform/kleidiai-explainer/page3.md:23]: Feels like Ubuntu is usually referred to with the full name, like Ubuntu 24.04. Would be nice to refer some AWS LP here as well

File: content/learning-paths/cross-platform/kleidiai-explainer/page2.md [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+
+## High-level KleidiAI architecture
+This section aims to provide an abstracted overview of KleidiAI's components before diving into the specifics. The KleidiAI source files are publicly accessible in the [KleidiAI GitLab repository](https://gitlab.arm.com/kleidi/kleidiai). Navigate there in your web browser to follow along and understand KleidiAI's structure.
+
+KlediAI's micro-kernels are located in the `/kai/ukernels/matmul` directory; navigate there now. There are essentially two types of KleidiAI micro-kernels:
+1. Quantizing/Packing routines    - under the `pack` directory.
+2. Matrix Multiplication routines - the three directories starting with `matmul_clamp`. Each directory contains routines specialized for a specific input data type.
+
+
+![KleidiAI stuff](KleidiAI-src.jpg "Figure 3. KleidiAI src directory")
```
Comment (annietllnd) [inline content/learning-paths/cross-platform/kleidiai-explainer/page2.md:17]: This is broken, I think the extension should be capitalized .JPG

File: content/learning-paths/cross-platform/kleidiai-explainer/page1.md [hunk 3]
```diff
@@ -0,0 +1,75 @@
+---
+title: KleidiAI and matrix multiplication
```
Comment (annietllnd) [inline content/learning-paths/cross-platform/kleidiai-explainer/page1.md:2]: This is a beautifully written overview! :100:

File: content/learning-paths/cross-platform/kleidiai-explainer/_index.md [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+    - Run a basic C++ matrix multiplication example to showcase the speedup the KleidiAI micro-kernels deliver.
+    
+prerequisites:
+    - An Arm Linux machine that implements the *i8mm* architecture feature; this example uses an AWS Graviton 3 instance.
+    - A basic understanding of linear algebra terminology such as dot product and matrix multiplication.
+
+author_primary: Zach Lasiuk
+### Tags
+skilllevels: Introductory 
+subjects: ML
```
Comment (annietllnd) [inline content/learning-paths/cross-platform/kleidiai-explainer/_index.md:20]: + GenAI, LLM, Coding, Neon?

File: content/learning-paths/cross-platform/kleidiai-explainer/_next-steps.md [hunk 5]
```diff
@@ -0,0 +1,24 @@
+---
+next_step_guidance: Check out the KleidiAI further reading to further understand what it is capable of. 
+
+recommended_path: /learning-paths/servers-and-cloud-computing/llama-cpu/
+
+further_reading:
```
Comment (annietllnd) [inline content/learning-paths/cross-platform/kleidiai-explainer/_next-steps.md:6]: Worth pointing to KleidiCV here?

**Review by annietllnd — COMMENTED** [review]
Great one Zach! I really like the overall-structure of it, so not much to add there. Found some smaller things, let's sync offline something is unclear

File: content/learning-paths/cross-platform/kleidiai-explainer/_index.md [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+
+who_is_this_for: This is an introductory topic for people wanting to learn how Generative AI workloads execute on hardware, and how KleidiAI accelerates it.
+
+learning_objectives: 
+    - Understand how basic math operations power Large Language Models.
+    - Learn how the KleidiAI micro-kernels speed up Generative AI inference performance.
+    - Run a basic C++ matrix multiplication example to showcase the speedup the KleidiAI micro-kernels deliver.
+    
+prerequisites:
+    - An Arm Linux machine that implements the *i8mm* architecture feature; this example uses an AWS Graviton 3 instance.
```
[the remark this answers is not in the document]
Comment (zachlasiuk) [reply content/learning-paths/cross-platform/kleidiai-explainer/_index.md]: Made edits inline based on your suggestions

File: content/learning-paths/cross-platform/kleidiai-explainer/page1.md [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+
+For example, in the image below, *z1* is calculated as a dot product of connected *x*'s and *w*'s from the previous layer. All *z* values in Layer 0 can therefore be efficiently calculated with a matrix multiplication operation.
+
+![Neural Network example#center](neural-node-pic.jpg "Figure 1. Zoomed in on neural network node")
+
+
+Sidebar:  In addition to *weights*, each neuron in a neural network is assigned a *bias*. These weights and biases are learned during training and make up a model's parameters. For example, in the Llama 3 model with 8 billion parameters, the model has around 8 billion individual weights and biases that embody what it learned during training. Generally speaking, the more parameters a model has, the more information it can retain from its training, leading to more capable models.
```
[the remark this answers is not in the document]
Comment (zachlasiuk) [reply content/learning-paths/cross-platform/kleidiai-explainer/page1.md:36]: Good idea, linked to its model card on HF

File: content/learning-paths/cross-platform/kleidiai-explainer/page2.md [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+weight: 3
+
+### FIXED, DO NOT MODIFY
+layout: learningpathall
+---
+
+## High-level KleidiAI architecture
+This section aims to provide an abstracted overview of KleidiAI's components before diving into the specifics. The KleidiAI source files are publicly accessible in the [KleidiAI GitLab repository](https://gitlab.arm.com/kleidi/kleidiai). Navigate there in your web browser to follow along and understand KleidiAI's structure.
+
+KlediAI's micro-kernels are located in the `/kai/ukernels/matmul` directory; navigate there now. There are essentially two types of KleidiAI micro-kernels:
```
[the remark this answers is not in the document]
Comment (zachlasiuk) [reply content/learning-paths/cross-platform/kleidiai-explainer/page2.md:12]: good idea, done

File: content/learning-paths/cross-platform/kleidiai-explainer/page2.md (hunk 2 above)
Comment (zachlasiuk) [reply content/learning-paths/cross-platform/kleidiai-explainer/page2.md:17]: updated to capital

File: content/learning-paths/cross-platform/kleidiai-explainer/page2.md [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+Before deep-diving into KleidiAI's code, it is helpful to see how KleidiAI micro-kernels interact with an GenAI model and ML Framework at a high-level. The steps below describe in words how KleidiAI speeds up matrix multiplication. The example is of a user asking a chatbot app a question on their smartphone. This is the example application stack to analyze:
+
+![KleidiAI in Stack](sw-stack.png "KleidiAI in a real-world software stack.")
+
+### Simple inference walkthrough
+
+#### Stage 1: Input
+* **A user** inputs their question into the chatbot, such as "What is the capital of the United States?"
+* **The chatbot app** uses MediaPipe to convert that text into a series of tokens representing the question as a matrix of FP16 numbers (as the Gemma-2b model is in FP16 format). 
+* **MediaPipe** invokes the LLM inference with the tokenized input, feeding it into the first neural network layer of the Gemma-2b model.
```
[the remark this answers is not in the document]
Comment (zachlasiuk) [reply content/learning-paths/cross-platform/kleidiai-explainer/page2.md:44]: Good idea, done

File: content/learning-paths/cross-platform/kleidiai-explainer/page2.md [hunk 10]
```diff
@@ -0,0 +1,96 @@
...
+
+As a result, KleidiAI strategically selected INT4 quantization for model parameters and INT8 for inputs propagating through the neural network. 
+
+{{% notice What AI model size should you select %}}
+Most models are trained in FP32 or FP16 formats and are by default available to download at that size. To realize the benefits of lower memory footprint, select a pre-quantized version of your desired AI model, ideally in INT4 format. You can quantize models yourself through various tools, but the quickest and easiest way is to locate a pre-quantized version of the same model.
+
+KleidiAI supports matrix multiplication of models quantized to FP32, FP16, INT8, and INT4 - your ML framework provider will select the optimal KleidiAI micro-kernel to balance inference speed and accuracy for your use-case.
```
[the remark this answers is not in the document]
Comment (zachlasiuk) [reply content/learning-paths/cross-platform/kleidiai-explainer/page2.md:82]: Great point, I modified language to make clearer

File: content/learning-paths/cross-platform/kleidiai-explainer/page3.md (hunk 1 above)
Comment (zachlasiuk) [reply content/learning-paths/cross-platform/kleidiai-explainer/page3.md:23]: Done for both

File: content/learning-paths/cross-platform/kleidiai-explainer/_next-steps.md (hunk 5 above)
Comment (zachlasiuk) [reply content/learning-paths/cross-platform/kleidiai-explainer/_next-steps.md:6]: I personally don't think so as that is for a different set of users doing graphics. KleidiAI is focusing on LLM-based devs

File: content/learning-paths/cross-platform/kleidiai-explainer/_index.md (hunk 4 above)
Comment (zachlasiuk) [reply content/learning-paths/cross-platform/kleidiai-explainer/_index.md:20]: good call, added to 'tools_software_languages'

File: content/learning-paths/cross-platform/kleidiai-explainer/_next-steps.md (hunk 5 above)
Comment (annietllnd) [reply content/learning-paths/cross-platform/kleidiai-explainer/_next-steps.md:6]: No worries, let's save the excitement for another LP

**Review by annietllnd — COMMENTED** [review]
Thanks for following up, looks good to me now.

**Pull request merged**